contracts/competition/Competition.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+import "../referrals/interfaces/IReferralStorage.sol";
+import "../access/Governable.sol";
+
+contract Competition is Governable
+{
+    struct Team {
+        address leader;
+        string name;
+        bytes32 referralCode;
+        address[] members;
+    }
+
+    struct Competition {
+        uint start;
+        uint end;
+        uint maxTeamSize;
+        mapping(address => Team) teams;
+        mapping(string => bool) teamNames;
+        mapping(address => address) memberTeams;
+        mapping(address => address) joinRequests;
+    }
+
+    uint public nextCompetitionIndex = 0;
+    mapping(uint => Competition) public competitions;
```

Comment:
i'm thinking if it's better to just use array instead of mapping, no need to manually manage index 🤔